### PR TITLE
Cs3 share manager

### PR DESCRIPTION
**Pull request by aduffeck**
This PR introduces a new share manager which uses the new metadata storage backend.

### CODE REVIEW

File: pkg/share/manager/cs3/cs3.go [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+type ReceivedShareMetadata struct {
+	State      collaboration.ShareState `json:"state"`
+	MountPoint *provider.Reference      `json:"mountpoint"`
+}
+
+func init() {
+	registry.Register("cs3", NewDefault)
+}
+
+type config struct {
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:82]: Can we make this public? Having an exported overview of the configuration for this might be beneficial

File: pkg/share/manager/cs3/cs3.go [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+func New(s Storage, indexer Indexer) (*Manager, error) {
+	return &Manager{
+		storage:     s,
+		indexer:     indexer,
+		initialized: false,
+	}, nil
+}
+
+func (m *Manager) initialize() error {
+	if m.initialized {
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:117]: We can have an concurrency issue here. Imagine two routines go into this function at the same the time. `m.initialized` would be false for both, so both start the initialization process. This can lead to unforeseeable consequences

My best approach was to `Lock` if not initialized and then check for `initialized` again. But there might be better ones.

File: pkg/share/manager/cs3/cs3.go [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+}
+
+// Share creates a new share
+func (m *Manager) Share(ctx context.Context, md *provider.ResourceInfo, g *collaboration.ShareGrant) (*collaboration.Share, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+	user := ctxpkg.ContextMustGetUser(ctx)
+	now := time.Now().UnixNano()
+	ts := &typespb.Timestamp{
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:156]: maybe use `utils.TSNow()` instead?

File: pkg/share/manager/cs3/cs3.go [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+	}
+
+	metadataPath := path.Join("metadata", share.Id.OpaqueId)
+	err = m.storage.MakeDirIfNotExist(ctx, metadataPath)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = m.indexer.Add(share)
+	if err != nil {
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:192]: `err` check is unnecessary. Just `return share, err`

File: pkg/share/manager/cs3/cs3.go [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+		return err
+	}
+
+	fn := path.Join("shares", ref.GetId().OpaqueId)
+	err = m.storage.Delete(ctx, fn)
+	if err != nil {
+		return err
+	}
+
+	return m.indexer.Delete(share)
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:231]: Are we sure we want to error if the indexer fails to delete but the storage was successful? What should the client do with that information?

File: pkg/share/manager/cs3/cs3.go [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+}
+
+// ListShares returns the shares created by the user. If md is provided is not nil,
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:234]: Comment unclear. What is `md`?

File: pkg/share/manager/cs3/cs3.go [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+	share.Permissions = p
+
+	data, err := json.Marshal(share)
+	if err != nil {
+		return nil, err
+	}
+
+	fn := path.Join("shares", share.Id.OpaqueId)
+	err = m.storage.SimpleUpload(ctx, fn, data)
+	if err != nil {
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:281]: Again unnecessary `if` statement

File: pkg/share/manager/cs3/cs3.go [hunk 8]
```diff
@@ -0,0 +1,564 @@
...
+	}
+
+	for _, id := range receivedIds {
+		share, err := m.getShareByID(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		metadata, err := m.downloadMetadata(ctx, share)
+		if err != nil {
+			if _, ok := err.(errtypes.NotFound); ok {
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:334]: Can we reverse the logic here?
```golang
if !ok {
    return nil, err
}
```
Then we can get rid of the else statement which reduces cognitive complexity.

File: pkg/share/manager/cs3/cs3.go [hunk 9]
```diff
@@ -0,0 +1,564 @@
...
+	}
+
+	metadata, err := m.downloadMetadata(ctx, share)
+	if err != nil {
+		if _, ok := err.(errtypes.NotFound); ok {
+			// use default values if the grantee didn't configure anything yet
+			metadata = ReceivedShareMetadata{
+				State: collaboration.ShareState_SHARE_STATE_PENDING,
+			}
+		} else {
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:370]: Same here. The `else` statement makes the function unnecessary complicated.

File: pkg/share/manager/cs3/cs3.go [hunk 10]
```diff
@@ -0,0 +1,564 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	fn, err := metadataFilename(rshare.Share, user)
+	if err != nil {
+		return nil, err
+	}
+	err = m.storage.SimpleUpload(ctx, fn, data)
+	if err != nil {
```
Comment (kobergj) [inline pkg/share/manager/cs3/cs3.go:424]: `err` check also here unnecessary

File: pkg/storage/utils/indexer/errors/errors.go [hunk 11]
```diff
@@ -50,8 +47,5 @@ func (e *NotFoundErr) Error() string {
 	return fmt.Sprintf("%s with %s=%s not found", e.TypeName, e.IndexBy.String(), e.Value)
 }
 
-// IsNotFoundErr checks whether an error is of type IsNotFoundErr.
-func IsNotFoundErr(e error) bool {
-	_, ok := e.(*NotFoundErr)
-	return ok
-}
+// IsNotFound implements the IsNotFound interface.
+func (e *NotFoundErr) IsNotFound() {}
```
Comment (kobergj) [inline pkg/storage/utils/indexer/errors/errors.go:51]: I like :+1: That's much simpler than before.

**Review by kobergj — CHANGES_REQUESTED** [review]
Found some smaller ones. Looks good in general!

File: pkg/share/manager/cs3/cs3.go [hunk 12]
```diff
@@ -0,0 +1,564 @@
...
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"google.golang.org/genproto/protobuf/field_mask"
+)
+
+//go:generate mockery -name Storage
+//go:generate mockery -name Indexer
+
+// Storage is the interface to the metadata storage backend
+type Storage interface {
```
[the remark this answers is not in the document]
Comment (aduffeck) [reply pkg/share/manager/cs3/cs3.go:52]: Yes, you are right. I'm currently refactoring this as part of the publicshare manager pull request here: https://github.com/aduffeck/reva/commit/f1b31ef53f9d7156b88b415e2f23a788c33fb52b 
So I hope it's fine to leave it in for this pr, it will be gone soon ;)

File: pkg/share/manager/cs3/cs3.go (hunk 1 above)
Comment (aduffeck) [reply pkg/share/manager/cs3/cs3.go:82]: Hm, maybe. But currently the struct is only used for extracting the configuration values from the generic interface which is being passed as part of the registry contract. The values are only passed on to the storage and not even stored in the manager itself.
Personally I would rather keep it private and expose it when it's actually being used.

File: pkg/share/manager/cs3/cs3.go (hunk 2 above)
Comment (aduffeck) [reply pkg/share/manager/cs3/cs3.go:117]: My idea was that running the initializing steps multiple times should be fine because they should be idempotent. I also *think* that running them in parallel shouldn't do any harm, but I agree it's better to err on the safe side here. I'll add some locking to it.

File: pkg/share/manager/cs3/cs3.go (hunk 4 above)
Comment (aduffeck) [reply pkg/share/manager/cs3/cs3.go:186]: But if the share couldn't be persisted in the index we don't want the caller to use the share object as that could lead to inconsistencies. I think it's safer to not give that option in the first place, the share hasn't actually be (fully) created.

File: pkg/share/manager/cs3/cs3.go (hunk 7 above)
Comment (aduffeck) [reply pkg/share/manager/cs3/cs3.go:277]: See my comment above.

File: pkg/share/manager/cs3/cs3.go (hunk 4 above)
Comment (kobergj) [reply pkg/share/manager/cs3/cs3.go:186]: We return a non-nil error. That should be enough indication that the share is not fully initialized and you shouldn't use it at all. I think it is common understanding in go that you do not use a value if the corresponding error is not nil. 

So in my opinion we can safely assume that the `share` value is not used if `err != nil`

Comment (aduffeck) [reply pkg/share/manager/cs3/cs3.go:196]: You are probably right and it'll never be a problem, my personal preference would still have been to keep it a tad cleaner at the cost of an additional `if`. I've also seen both ways being used in go projects. It's fine though, I'll change it and push a commit in a sec.

File: pkg/share/manager/cs3/cs3.go [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+		Ctime:       ts,
+		Mtime:       ts,
+	}
+
+	data, err := json.Marshal(share)
+	if err != nil {
+		return nil, err
+	}
+
+	fn := path.Join("shares", share.Id.OpaqueId)
```
Comment (dragotin) [inline pkg/share/manager/cs3/cs3.go:183]: I saw a few places where we hardcode path segments like this. I wonder if we should not have a central place where we define these instead of hardcoding.

Maybe not really part of this PR..

Comment (aduffeck) [reply pkg/share/manager/cs3/cs3.go]: @dragotin I just pushed a commit which changes the remaining places to also make use of `shareFilename()` which encapsulates the knowledge about the directoy layout.
Not sure if that's what you meant, but I don't think we need to move this to a central place inside the repository as the layout is only relevant to the share manager itself.

**Review by kobergj — APPROVED** [review]
